syntax = "proto3";

import "edge_parameter.proto";
import "vertex_parameter.proto";

package intellgraph;

message SolverConfig {
  // Required
  string type = 1;

  // Required
  float eta = 2;

  // Required
  float lambda = 3;
}

message GraphParameter {
  // Optional
  SolverConfig solver_config = 1;

  // Required, length represents the batch size for classical NN, and the 
  // sequence length for RNN
  int32 length = 2;

  // Required
  VertexParameter input_vertex_param = 3;

  // Required
  VertexParameter output_vertex_param = 4;

  // Optional
  repeated VertexParameter intermediate_vertex_params = 5;

  // Required
  repeated EdgeParameter edge_params = 6;

  // Optional, required for RNN
  map<int32, int32> state_vertex_map = 7;
}
